File: app/src/main/java/com/machiav3lli/backup/handler/action/BackupSpecialAction.java
package com.machiav3lli.backup.handler.action;

import android.content.Context;
import android.util.Log;

import com.machiav3lli.backup.Constants;
import com.machiav3lli.backup.handler.Crypto;
import com.machiav3lli.backup.handler.ShellHandler;
import com.machiav3lli.backup.items.ActionResult;
import com.machiav3lli.backup.items.AppInfo;
import com.machiav3lli.backup.utils.PrefUtils;

import org.apache.commons.io.FileUtils;

import java.io.File;
import java.io.IOException;
import java.util.Arrays;
import java.util.stream.Collectors;

public class BackupSpecialAction extends BackupAppAction {
    private static final String TAG = Constants.classTag(".BackupSpecialAction");

    public BackupSpecialAction(Context context, ShellHandler shell) {
        super(context, shell);
    }

    @Override
    public ActionResult run(AppInfo app, int backupMode) {
        if ((backupMode & AppInfo.MODE_APK) == AppInfo.MODE_APK) {
            Log.e(BackupSpecialAction.TAG, String.format("%s", "Special contents don't have APKs to backup. Ignoring"));
        }
        if ((backupMode & AppInfo.MODE_DATA) == AppInfo.MODE_DATA) {
            return super.run(app, AppInfo.MODE_DATA);
        }
        return new ActionResult(app, "Special backup only backups data, but data was not selected for backup", false);
    }

    @Override
    protected void backupData(AppInfo app) throws BackupFailedException, Crypto.CryptoSetupException {
        Log.i(BackupSpecialAction.TAG, String.format("%s: Backup special data", app));
        File backupDirectory = this.getDataBackupFolder(app);
        // Create the (temporary) directory for the backup data
        if (!(backupDirectory.exists() || backupDirectory.mkdir())) {
            String errorMessage = String.format("%s: Could not create temporary backup directory: %s", app, backupDirectory);
            Log.e(BackupSpecialAction.TAG, String.format("%s: %s", app, errorMessage));
            throw new BackupFailedException(errorMessage, null);
        }

        String command = this.prependUtilbox(String.format(
                "cp -RLp %s %s",
                Arrays.stream(app.getFilesList()).map(s -> '"' + s + '"').collect(Collectors.joining(" ")),
                backupDirectory
        ));
        try {
            ShellHandler.runAsRoot(command);
            this.compress(
                    backupDirectory,
                    this.getBackupArchive(app, BaseAppAction.BACKUP_DIR_DATA, PrefUtils.isEncryptionEnabled(this.getContext()))
            );
        } catch (ShellHandler.ShellCommandFailedException e) {
            String error = BaseAppAction.extractErrorMessage(e.getShellResult());
            Log.e(BackupSpecialAction.TAG, String.format("%s: Backup Special Data failed: %s", app, error));
            throw new BackupFailedException(error, e);
        } catch (IOException e) {
            Log.e(BackupSpecialAction.TAG, String.format("%s: Backup Special Data failed with IOException: %s", app, e));
            throw new BackupFailedException("IOException", e);
        } finally {
            // if the backup is compressed, clean up in any case
            boolean backupDeleted = FileUtils.deleteQuietly(backupDirectory);
            Log.d(BackupSpecialAction.TAG, String.format("%s: Uncompressed Backup was deleted: %s", app, backupDeleted));
        }
    }

    @Override
    public boolean cleanBackup(AppInfo app, int backupMode) {
        if ((backupMode & AppInfo.MODE_APK) == AppInfo.MODE_APK) {
            Log.w(BackupSpecialAction.TAG, String.format("%s: Refusing to clean package data of a special app with mode MODE_APK.", app));
        } else if ((backupMode & AppInfo.MODE_DATA) == AppInfo.MODE_DATA) {
            return super.cleanBackup(app, AppInfo.MODE_DATA);
        }
        return false;
    }

    // Stubbing some functions, to avoid executing them with potentially dangerous results
    @Override
    protected void backupPackage(AppInfo app) {
        // stub
    }

    @Override
    protected void backupDeviceProtectedData(AppInfo app) {
        // stub
    }

    @Override
    protected void backupExternalData(AppInfo app) {
        // stub
    }

    @Override
    protected void backupObbData(AppInfo app) {
        // stub
    }
}

File: app/src/main/java/com/machiav3lli/backup/handler/action/BackupSpecialAction.kt
/*
 * OAndBackupX: open-source apps backup and restore app.
 * Copyright (C) 2020  Antonios Hazim
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.machiav3lli.backup.handler.action

import android.content.Context
import android.util.Log
import com.machiav3lli.backup.Constants.classTag
import com.machiav3lli.backup.handler.Crypto.CryptoSetupException
import com.machiav3lli.backup.handler.ShellHandler
import com.machiav3lli.backup.handler.ShellHandler.ShellCommandFailedException
import com.machiav3lli.backup.items.ActionResult
import com.machiav3lli.backup.items.AppInfoX
import com.machiav3lli.backup.items.SpecialAppMetaInfo
import com.machiav3lli.backup.items.StorageFile
import java.io.File

class BackupSpecialAction(context: Context, shell: ShellHandler) : BackupAppAction(context, shell) {

    override fun run(app: AppInfoX, backupMode: Int): ActionResult? {
        if (backupMode and MODE_APK == MODE_APK) {
            Log.e(TAG, "Special contents don't have APKs to backup. Ignoring")
        }
        return if (backupMode and MODE_DATA == MODE_DATA) super.run(app, MODE_DATA)
        else ActionResult(app, null,
                "Special backup only backups data, but data was not selected for backup", false)
    }

    @Throws(BackupFailedException::class, CryptoSetupException::class)
    override fun backupData(app: AppInfoX, backupInstanceDir: StorageFile?): Boolean {
        Log.i(TAG, "$app: Backup special data")
        require(app.appInfo is SpecialAppMetaInfo) { "Provided app is not an instance of SpecialAppMetaInfo" }
        val appInfo = app.appInfo as SpecialAppMetaInfo?
        // Get file list
        // This can be optimized, because it's known, that special backups won't meet any symlinks
        // since the list of files is fixed
        // It would make sense to implement something like TarUtils.addFilepath with SuFileInputStream and
        val filesToBackup: MutableList<ShellHandler.FileInfo> = ArrayList(appInfo!!.fileList.size)
        try {
            for (filepath in appInfo.fileList) {
                val isDirSource = filepath!!.endsWith("/")
                val parent = if (isDirSource) File(filepath).name else null
                val fileInfos = shell.suGetDetailedDirectoryContents(filepath, false, parent)
                if (isDirSource) {
                    filesToBackup.add(ShellHandler.FileInfo(parent!!, ShellHandler.FileInfo.FileType.DIRECTORY,
                            File(filepath).parent!!, "system", "system", 504.toShort(), 0))
                }
                filesToBackup.addAll(fileInfos)
            }
            genericBackupData(BACKUP_DIR_DATA, backupInstanceDir!!.uri, filesToBackup, true)
        } catch (e: ShellCommandFailedException) {
            val error = extractErrorMessage(e.shellResult)
            Log.e(TAG, "$app: Backup Special Data failed: $error")
            throw BackupFailedException(error, e)
        }
        return true
    }

    // Stubbing some functions, to avoid executing them with potentially dangerous results
    override fun backupPackage(app: AppInfoX, backupInstanceDir: StorageFile?) {
        // stub
    }

    override fun backupDeviceProtectedData(app: AppInfoX, backupInstanceDir: StorageFile?): Boolean {
        // stub
        return false
    }

    override fun backupExternalData(app: AppInfoX, backupInstanceDir: StorageFile?): Boolean {
        // stub
        return false
    }

    override fun backupObbData(app: AppInfoX, backupInstanceDir: StorageFile?): Boolean {
        // stub
        return false
    }

    override fun preprocessPackage(packageName: String) {
        // stub
    }

    override fun postprocessPackage(packageName: String) {
        // stub
    }

    companion object {
        private val TAG = classTag(".BackupSpecialAction")
    }
}